util/eventbus/bus.go

```diff
@@ -0,0 +1,216 @@
+		// selecting on close/snapshot, those can get handled on the
+		// next dispatch cycle.
+	unblockPublishers:
+		for {
+			select {
+			case val := <-acceptCh():
+				vals.Add(val)
+			default:
+				break unblockPublishers
+			}
+		}
+	}
+}
+
+func (b *Bus) dest(t reflect.Type) []*Queue {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	return b.topics[t]
+}
+
+func (b *Bus) subscribe(t reflect.Type, q *Queue) (cancel func()) {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	b.topics[t] = append(b.topics[t], q)
+	return func() {
+		b.unsubscribe(t, q)
+	}
+}
+
+func (b *Bus) unsubscribe(t reflect.Type, q *Queue) {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	// Topic slices are accessed by pump without holding a lock, so we
+	// have to replace the entire slice when unsubscribing.
+	// Unsubscribing should be infrequent enough that this won't
+	// matter.
+	i := slices.Index(b.topics[t], q)
+	if i == -1 {
+		return
+	}
+	b.topics[t] = slices.Delete(slices.Clone(b.topics[t]), i, i+1)
+}
+
+func (b *Bus) Close() {
+	b.stop.Close()
+}
+
+// Queue returns a new queue with no subscriptions. Use [Subscribe] to
+// atach subscriptions to it.
+//
+// The queue's name should be a short, human-readable string that
+// identifies this queue. The name is only visible through debugging
+// APIs.
+func (b *Bus) Queue(name string) *Queue {
+	return newQueue(b, name)
+}
+
+func (b *Bus) addQueue(q *Queue) {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	b.queues.Add(q)
+}
+
+func (b *Bus) deleteQueue(q *Queue) {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	b.queues.Delete(q)
+}
+
+func (b *Bus) addPublisher(p publisher) {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	b.publishers.Add(p)
+}
+
+func (b *Bus) deletePublisher(p publisher) {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	b.publishers.Delete(p)
+}
+
+// goroutineShutdown is a helper type to signal a worker goroutine to
+// shut down, and wait until it has finished shutting down.
+type goroutineShutdown struct {
+	ctx      context.Context
+	cancel   context.CancelFunc
+	finished chan struct{}
+}
+
+func newGoroutineShutdown() *goroutineShutdown {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &goroutineShutdown{
+		ctx:      ctx,
+		cancel:   cancel,
+		finished: make(chan struct{}),
+	}
+}
+
+// Close signals a worker goroutine to shut down, and waits for it to
+// report that shutdown is complete.
+func (g *goroutineShutdown) Close() {
+	g.cancel()
+	<-g.finished
+}
+
+// Closed returns a channel that's closed when the worker goroutine
+// has shut down.
+func (g *goroutineShutdown) Closed() <-chan struct{} {
+	return g.finished
+}
+
+// StartShutdown returns a channel that's closed when the worker
+// goroutine should start shutdown. The worker goroutine should call
+// ShutdownComplete immediately prior to exiting.
+func (g *goroutineShutdown) StartShutdown() <-chan struct{} {
+	return g.ctx.Done()
+}
+
+func (g *goroutineShutdown) ShutdownComplete() {
```

Comment:
The name here confused me at first, I get it now, but what do you think about 

```
Stop()     // tell the goroutine to stop
Wait()     // wait for the goroutine to finish
StopWait() // stop + wait
```

as an API surface here?

(I suggested "Stop" rather than "Close" because it doesn't have the signature of io.Closer, which is fine but maybe distracting? YMMV though)

Reply:
Yeah I'm not quite happy with the API shape here, which is complicated due to having 2 different facets: one API for controlling the shutdown, and another for responding to it. And they both would like to use the same vocabulary.

I made a tentative change and split this helper into 2 values, to make the "for the controller" and "for the worker" split explicit. WDYT?

Reply:
Hmm, yeah, my worry about splitting the vocab like that is that there's no "obvious" split between which methods are meant for the worker vs. the controller, and it trips up my flow slightly to have to check it on use. But that may improve with use, and I don't really like the two-object split on reflection. I'll send a followup that reworks to what you suggested, and we can revisit down the road and see if we ended up liking it.